### PR TITLE
[task][golang-test] add support for private dependencies and cgo enabled source

### PR DESCRIPTION
**Pull request by ProgramCpp**
# Changes
adds support to configure go env variables to,
    - download private module dependencies for tests. this includes configuring ssh keys and adding git config to use ssh to download dependencies
    - build C integrated code by enabling cgo. this includes installing C dependencies from apt.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Follows the [authoring recommendations](https://github.com/tektoncd/catalog/blob/main/recommendations.md)
- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [X] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [X] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [X] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [X] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [X] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

---

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

### CODE REVIEW

**Comment by keskad** [comment]
How about adding a possibility to parametrize also the image? I'm thinking about air-gapped environments.

**Comment by keskad** [comment]
Would it be possible to add GOPROXY environment variable?
https://jfrog.com/blog/why-goproxy-matters-and-which-to-pick/

**Comment by ProgramCpp** [comment]
- added GOPROXY variable
- parameterized docker image to run the task

File: task/golang-test/0.3/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+* **GO111MODULE**: value of module support (_default:_ auto)
+* **GOCACHE**: value for go caching path (_default:_ "")
+* **GOMODCACHE**: value for go module caching path (_default:_ "")
+* **CGOENABLED**: go env. enables cgo
+* **GOPRIVATE**: go env. private modules to be fetched from original source
+* **GIT_ALLOW_PROTOCOL**: list of vcs schemes to download module dependencies
+* **git_config_add**: git configuration to add
+* **pkgs**: linux packages to install from the apt library
+* **pkg_source**: additional linux package sources to fetch apt packages from
+
```
Comment (keskad) [inline task/golang-test/0.3/README.md:29]: Missing:

`* **GOPROXY**: [Proxy settings](https://go.dev/ref/mod#goproxy-protocol) for Go dependencies (default: empty, not enabled)`

:wink:

File: task/golang-test/0.3/golang-test.yaml [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    - name: GOCACHE
+      value: "$(params.GOCACHE)"
+    - name: GOMODCACHE
+      value: "$(params.GOMODCACHE)"
+    - name: GOPRIVATE
+      value: "$(params.GOPRIVATE)"
+    - name: CGOENABLED
+      value: "$(params.CGOENABLED)"
+    - name: GIT_ALLOW_PROTOCOL
+      value: "$(params.GIT_ALLOW_PROTOCOL)"
```
Comment (keskad) [inline task/golang-test/0.3/golang-test.yaml:131]: `GOPROXY` was added as param, but not used anywhere. I guess it should be passed there as environment variable? :yum:

File: task/golang-test/0.3/golang-test.yaml [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    description: "Go caching directory path"
+    default: ""
+  - name: GOMODCACHE
+    description: "Go mod caching directory path"
+    default: ""
+  - name: CGOENABLED
+    description: enable cgo 
+  - name: GOPRIVATE
+    description: "list of go private modules"
+  - name: GOPROXY
```
Comment (keskad) [inline task/golang-test/0.3/golang-test.yaml:54]: According to https://go.dev/ref/mod#goproxy-protocol the default value should be "off" or "direct"

File: task/golang-test/0.3/README.md (hunk 1 above)
Comment (ProgramCpp) [reply task/golang-test/0.3/README.md:29]: thanks for pointing it out 👍

File: task/golang-test/0.3/golang-test.yaml (hunk 2 above)
Comment (ProgramCpp) [reply task/golang-test/0.3/golang-test.yaml:131]: oops, sorry

File: task/golang-test/0.3/golang-test.yaml (hunk 3 above)
Comment (ProgramCpp) [reply task/golang-test/0.3/golang-test.yaml:54]: off/ direct are valid values. proxy is enabled by default. Ive added default values from my unedited environment

**Review by vdemeester — APPROVED** [review]
Thanks 🎉 
Can we have this for the `golang-build` task as well ?

**Review by keskad — APPROVED** [review]

**Comment by ProgramCpp** [comment]
@vdemeester can i create a separate PR for the build task?

**Comment by vdemeester** [comment]
> @vdemeester can i create a separate PR for the build task?

Of course 😉

**Comment by ProgramCpp** [comment]
@vdemeester could you help merge this PR? this pr requires a lgtm label

File: task/golang-test/0.3/README.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Golang Test
+
+This task is a Golang task to test Go projects.
+
+## Install the task
+
+```
+kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/main/task/golang-test/0.2/golang-test.yaml
```
Comment (vinamra28) [inline task/golang-test/0.3/README.md:8]:
```suggestion
kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/main/task/golang-test/0.3/golang-test.yaml
```

File: task/golang-test/0.3/golang-test.yaml [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    default: "./..."
+  - name: context
+    description: path to the directory to use as context.
+    default: "."
+  - name: image
+    description: docker image to run the task
+    default: "docker.io/library/golang"
+  - name: version
+    description: golang version to use for tests
+    default: "latest"
```
Comment (vinamra28) [inline task/golang-test/0.3/golang-test.yaml:31]: instead of `latest` can we tag to a particular version so that in future tests don't fail?

**Review by vinamra28 — COMMENTED** [review]
/ok-to-test

File: task/golang-test/0.3/golang-test.yaml [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  - name: GOCACHE
+    description: "Go caching directory path"
+    default: ""
+  - name: GOMODCACHE
+    description: "Go mod caching directory path"
+    default: ""
+  - name: CGOENABLED
+    description: enable cgo 
+  - name: GOPRIVATE
+    description: "list of go private modules"
```
Comment (vinamra28) [inline task/golang-test/0.3/golang-test.yaml:53]: shouldn't these have default values defined? 🤔

File: task/golang-test/0.3/golang-test.yaml [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    description: whitelist go module checksum DB's
+    default: "sum.golang.org"
+  - name: GIT_ALLOW_PROTOCOL
+    description: "list of colon separated vcs schemes"
+  - name: git_config_add
+    description: git config to add in space separated format. <name> <value>
+  - name: pkgs
+    description: linux packages to install from the apt library in space separated format. <pkg1> <pkg2>
+  - name: pkg_source
+    description: additional linux package sources to fetch apt packages from
```
Comment (vinamra28) [inline task/golang-test/0.3/golang-test.yaml:67]: are these mandatory parameters or can these have default values?

**Comment by chmouel** [comment]
I am not sure I am very keen to have this task doing package installation, i think that goes beyond customization a task should offer... 

what about making the image customization and let the user do the package install to a new image, instead?

**Comment by ProgramCpp** [comment]
> agree @chmouel 

since package installation is optional, is it okay to have this option open? helps to setup/ test pipelines locally without any container registry dependency

**Pull request closed without merging**